src/hotspot/share/runtime/deoptimization.cpp

```diff
@@ -934,113 +936,90 @@ JRT_LEAF(BasicType, Deoptimization::unpack_frames(JavaThread* thread, int exec_m
     rm.set_include_argument_oops(false);
     bool is_top_frame = true;
     int callee_size_of_parameters = 0;
-    int callee_max_locals = 0;
-    for (int i = 0; i < cur_array->frames(); i++) {
-      vframeArrayElement* el = cur_array->element(i);
+    for (int frame_idx = 0; frame_idx < cur_array->frames(); frame_idx++) {
+      assert(is_top_frame == (frame_idx == 0), "must be");
+      vframeArrayElement* el = cur_array->element(frame_idx);
       frame* iframe = el->iframe();
       guarantee(iframe->is_interpreted_frame(), "Wrong frame type");
+      methodHandle mh(thread, iframe->interpreter_frame_method());
+      bool reexecute = el->should_reexecute();
 
-      // Get the oop map for this bci
-      InterpreterOopMap mask;
       int cur_invoke_parameter_size = 0;
-      bool try_next_mask = false;
-      int next_mask_expression_stack_size = -1;
       int top_frame_expression_stack_adjustment = 0;
-      methodHandle mh(thread, iframe->interpreter_frame_method());
-      OopMapCache::compute_one_oop_map(mh, iframe->interpreter_frame_bci(), &mask);
-      BytecodeStream str(mh, iframe->interpreter_frame_bci());
       int max_bci = mh->code_size();
-      // Get to the next bytecode if possible
+      BytecodeStream str(mh, iframe->interpreter_frame_bci());
+      assert(str.bci() < max_bci, "bci in interpreter frame out of bounds");
+      Bytecodes::Code cur_code = str.next();
+
+      if (!reexecute && !Bytecodes::is_invoke(cur_code)) {
+        assert(is_top_frame, "must be");
+        assert(falls_through(cur_code), "must be");
+        assert(cur_code != Bytecodes::_illegal, "illegal bytecode");
+        assert(str.bci() < max_bci, "bci in interpreter frame out of bounds");
+
+        // Need to subtract off the size of the result type of
+        // the bytecode because this is not described in the
+        // debug info but returned to the interpreter in the TOS
+        // caching register
+        BasicType bytecode_result_type = Bytecodes::result_type(cur_code);
+        if (bytecode_result_type != T_ILLEGAL) {
+          top_frame_expression_stack_adjustment = type2size[bytecode_result_type];
+        }
+        assert(top_frame_expression_stack_adjustment >= 0, "stack adjustment must be positive");
+
+        cur_code = str.next();
+        reexecute = true;
```

Comment:
This seems a little unsavory, particularly since there's a later step which will print that value as if it was the original one.  Since there's only one later logic use of the variable maybe there's should be a new flag to mark special case?  Like `rolled_forward`?  It might be fine as is with comments explaining this and fixing the printing to reflect what occurred here.